### PR TITLE
Laboratory work #1, Timur Menyalshchikov - 21FPL1

### PR DESCRIPTION
**Pull request by Timur20033**

### CODE REVIEW

File: lab_1_keywords_tfidf/main.py [hunk 1]
```diff
@@ -17,7 +19,21 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(text, str):
+        text_without_punctuation = ""
+        for i in text:
+            if i.isalpha():
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:25]: This is not a check 'is not punctuation' - this check about 'is alpha and is digit and is a space'

File: lab_1_keywords_tfidf/start.py [hunk 2]
```diff
@@ -34,3 +34,6 @@
     RESULT = None
     # DO NOT REMOVE NEXT LINE - KEEP IT INTENTIONALLY LAST
     assert RESULT, 'Keywords are not extracted'
+
+
+print("meow")
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/start.py:39]: please, remove it

File: lab_1_keywords_tfidf/main.py [hunk 3]
```diff
@@ -156,4 +276,7 @@ def extract_significant_words(chi_values: dict[str, float], alpha: float) -> Opt
 
     In case of corrupt input arguments, None is returned
     """
+
     pass
+
+
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:282]: remove this line

File: lab_1_keywords_tfidf/main.py [hunk 4]
```diff
@@ -82,28 +158,72 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
...
+                else:
+                    return None
+
+    else:
+        return None
 
 
 def calculate_expected_frequency(
-    doc_freqs: dict[str, int], corpus_freqs: dict[str, int]
+        doc_freqs: dict[str, int], corpus_freqs: dict[str, int]
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:226]: remove spaces

File: lab_1_keywords_tfidf/start.py (hunk 2 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/start.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 3 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 4 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py (hunk 1 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: Done

File: lab_1_keywords_tfidf/main.py [hunk 5]
```diff
@@ -17,7 +18,17 @@ def clean_and_tokenize(text: str) -> Optional[list[str]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(text, str):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:21]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 6]
```diff
@@ -33,7 +44,22 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(tokens, list) and isinstance(stop_words, list):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:47]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 7]
```diff
@@ -48,7 +74,14 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(tokens, list) and tokens != []:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:77]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 8]
```diff
@@ -48,7 +74,14 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
...
+        if all(isinstance(word, str) for word in tokens):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:78]: and this

File: lab_1_keywords_tfidf/main.py [hunk 9]
```diff
@@ -66,7 +99,26 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if (isinstance(frequencies, dict) and frequencies != {} and isinstance(top, int) and top > 0
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:102]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 10]
```diff
@@ -82,24 +134,49 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(frequencies, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:137]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 11]
```diff
@@ -82,24 +134,49 @@ def calculate_tf(frequencies: dict[str, int]) -> Optional[dict[str, float]]:
...
+       Parameters:
+       term_freq (Dict): A dictionary with tokens and its corresponding TF values
+       idf (Dict): A dictionary with tokens and its corresponding IDF values
+
+       Returns:
+       Dict: A dictionary with tokens and its corresponding TF-IDF values
+
+       In case of corrupt input arguments, None is returned
+       """
+    if isinstance(term_freq, dict) and isinstance(idf, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:164]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 12]
```diff
@@ -118,7 +195,40 @@ def calculate_expected_frequency(
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(doc_freqs, dict) and isinstance(corpus_freqs, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:198]: revert the condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 13]
```diff
@@ -137,7 +247,23 @@ def calculate_chi_values(expected: dict[str, float], observed: dict[str, int]) -
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(expected, dict) and isinstance(observed, dict):
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:250]: revert every condition and return None immediately

File: lab_1_keywords_tfidf/main.py [hunk 14]
```diff
@@ -156,4 +282,21 @@ def extract_significant_words(chi_values: dict[str, float], alpha: float) -> Opt
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if isinstance(chi_values, dict) and isinstance(alpha, float):
+        if (all(isinstance(word, str) for word in chi_values.keys())
+                and all(isinstance(freq, float) for freq in chi_values.values()) and chi_values != {}):
+            if alpha in [0.05, 0.01, 0.001]:
+                final_dict = {}
+                if alpha == 0.05:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:290]: can you create a dict for this?

**Review by artyomtugaryov — CHANGES_REQUESTED** [review]

File: lab_1_keywords_tfidf/main.py (hunk 14 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 13 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 12 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 10 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 9 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 8 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 7 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 6 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py (hunk 5 above)
Comment (Timur20033) [reply lab_1_keywords_tfidf/main.py]: done

File: lab_1_keywords_tfidf/main.py [hunk 15]
```diff
@@ -33,7 +43,22 @@ def remove_stop_words(tokens: list[str], stop_words: list[str]) -> Optional[list
...
+    if stop_words != []:
+        if not (all(isinstance(word, str) for word in tokens)
+                and all(isinstance(word, str) for word in stop_words)):
+            return None
+        final_text = []
+        for word3 in tokens:
+            if word3 not in stop_words:
+                final_text += [word3]
+        return final_text
+    final_text = []
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:57]: .

File: lab_1_keywords_tfidf/main.py [hunk 16]
```diff
@@ -48,7 +73,14 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
 
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not isinstance(tokens, list) or not tokens != []:
+        return None
+    if not all(isinstance(word, str) for word in tokens):
+        return None
+    frequency_dictionary = {}
+    for word in tokens:
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:81]: set

Comment (artyomtugaryov) [reply lab_1_keywords_tfidf/main.py:81]: for word in set(tokens)

File: lab_1_keywords_tfidf/main.py [hunk 17]
```diff
@@ -48,7 +73,14 @@ def calculate_frequencies(tokens: list[str]) -> Optional[dict[str, int]]:
...
+        frequency_dictionary.update({word: tokens.count(word)})
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:82]: frequency_dictionary[] =

File: lab_1_keywords_tfidf/main.py [hunk 18]
```diff
@@ -66,7 +98,26 @@ def get_top_n(frequencies: dict[str, Union[int, float]], top: int) -> Optional[l
...
     In case of corrupt input arguments, None is returned
     """
-    pass
+    if not (isinstance(frequencies, dict) and frequencies != {} and isinstance(top, int) and top > 0
+            and isinstance(top, bool) is False):
+        return None
+    if not (all(isinstance(word, str) for word in frequencies.keys())
+            and all(isinstance(freq, (int, float)) for freq in frequencies.values())):
+        return None
+    sort_dictionaries_lst = sorted(frequencies.values(), reverse=True)
```
Comment (artyomtugaryov) [inline lab_1_keywords_tfidf/main.py:107]: frequencies.values() - > frequencies.keys(), key = lambda x: freq[x]
slice -> [:top]

**Pull request closed without merging**